orangecontrib/text/preprocess.py

```diff
@@ -1,94 +1,298 @@
+            self.tokenizer = None   # TODO: Change when twitter is available.
+
+        # Lowercase.
         self.lowercase = lowercase
-        self.transformation = trans
+
+        # Stop words.
+        if stop_words == 'english':
+            self.stop_words = set(stopwords.words('english'))
+        elif isinstance(stop_words, list):
+            self.stop_words = set(stop_words)
+        elif stop_words is None:
+            self.stop_words = None
+        else:
+            raise ValueError('The stop words parameter should be either '
+                             '"english", a list containing the stop words '
+                             'or None.')
+
+        # Min/Max df.
+        # Constant checks whether or not we should use 'df', are quite painful.
+        self.use_df_sw = False  # This flag summarizes this check.
+        self._check_df_constraints(min_df, max_df)
         self.min_df = min_df
-        self.cv = CountVectorizer(
-            lowercase=lowercase,
-            stop_words=stop_words,
-            preprocessor=trans,
-            min_df=min_df,
-        )
+        self.max_df = max_df
+        self.df_stop_words = None
+
+        # Transformation.
+        if transformation is not None and not isinstance(transformation,
+                                                         Stemmatizer):
+            raise ValueError('Type "{}" not supported.'.format(transformation))
+        self.transformation = transformation
 
     def __call__(self, data):
+        """Calls the necessary methods to perform pre-processing on the input.
+
+        Args:
+            data (str, list, `orangecontrib.text.corpus.Corpus`): The input we
+                want to process.
+
+        Returns:
+            str, list: The pre-processed input in the form of either a string
+                or a list, depending on the input.
+
+        Raises:
+            ValueError: If the input is not of expected type.
+        """
+        if isinstance(data, Corpus):
+            data = data.documents
+
+        if isinstance(data, str):
+            pp_corpus = self._preprocess_document(data)
+            return pp_corpus
+        elif isinstance(data, list):
+            if self.use_df_sw:  # DF parameters were specified.
+                self._get_corpus_specific_stopwords(data)
+            pp_corpus = [self._preprocess_document(doc) for doc in data]
+            return pp_corpus
+        else:
+            raise ValueError('Type "{}" not supported.'.format(type(data)))
+
+    def _preprocess_document(self, data):
+        # Tokenize.
+        data = self._tokenize(data)
+        # Lowercase.
+        if self.lowercase:
+            data = [token.lower() for token in data]
+        # Remove stop words.
+        if self.stop_words is not None:
+            data = self._remove_stop_words(data, self.stop_words)
+        if self.df_stop_words is not None:
+            data = self._remove_stop_words(data, self.df_stop_words)
```

Comment:
Each `_remove_stop_words ` is a pass over the document. There is no need to scan the documents twice since we can union `self.stop_words` and `self.df_stop_words` and then have only one scan.